Save integral tensor data for ET

shengfukevin opened the pull request:
Summary:
et_replay uses random data to run operators, however, the operators using index tensor to access memory won't work with random data. It usually ran into two exceptions: 1. illegal memory access since index is out of range, it has been fixed with the environment variable ENABLE_PYTORCH_EXECUTION_TRACE_SAVE_INTEGRAL_TENSOR_RANGE to record the min/max value of index tensors. 2. unaligned memory access, FBGEMM ops have speical requirements for the memory layout.

To fix the second execption, ENABLE_PYTORCH_EXECUTION_TRACE_SAVE_INTEGRAL_TENSOR is added to allow user to specify the node names, separated by comma, so ET will save the integral tensor data for these nodes. The saved data will be used in et_replay.

Be careful to turn on this option since it will use more space to save the extra data.

Test Plan: buck2 run mode/opt caffe2/test:test_profiler_cuda -- profiler.test_execution_trace.TestExecutionTraceCUDA.test_execution_trace_record_integral_tensor_data_cuda

Differential Revision: D67989856

malfet commented on torch/csrc/profiler/standalone/execution_trace_observer.cpp:
```diff
@@ -377,9 +382,39 @@ static ExecutionTraceObserver::ID getObjectID(
+  if (fs.is_open()) {
+    if (device.is_cpu()) {
+      fs.write(
+          (const char*)storage_data_ptr + tensor_offset, (long)tensor_nbyte);
```
It's already a const char, isn't it?
```suggestion
          storage_data_ptr + tensor_offset, (long)tensor_nbyte);
```

shengfukevin replied:
fixed.

malfet commented on torch/csrc/profiler/standalone/execution_trace_observer.cpp:
```diff
@@ -377,9 +382,39 @@ static ExecutionTraceObserver::ID getObjectID(
   return iter->second;
 }
 
+void dumpTensorData2File(
```
It must be static...

shengfukevin replied:
fixed.

malfet commented on torch/csrc/profiler/standalone/execution_trace_observer.cpp:
```diff
@@ -377,9 +382,39 @@ static ExecutionTraceObserver::ID getObjectID(
+      fs.write(
+          (const char*)host_data.data() + tensor_offset, (long)tensor_nbyte);
+    }
+    fs.close();
```
Aren't fstream guaranteed to call close when destructor is called?

malfet commented on torch/csrc/profiler/standalone/execution_trace_observer.cpp:
```diff
@@ -377,9 +382,39 @@ static ExecutionTraceObserver::ID getObjectID(
+          c10::Device(c10::DeviceType::CPU),
+          false);
+      fs.write(
+          (const char*)host_data.data() + tensor_offset, (long)tensor_nbyte);
```
Why do you need cast there?
```suggestion
          host_data.data() + tensor_offset, (long)tensor_nbyte);
```

shengfukevin replied:
fixed.

malfet commented on torch/csrc/profiler/standalone/execution_trace_observer.cpp:
```diff
@@ -377,9 +382,39 @@ static ExecutionTraceObserver::ID getObjectID(
+          (const char*)storage_data_ptr + tensor_offset, (long)tensor_nbyte);
+    } else {
+      std::vector<char> host_data(tensor_nbyte);
+      c10::CopyBytes(
+          tensor_nbyte,
+          storage_data_ptr,
+          device,
+          (void*)host_data.data(),
+          c10::Device(c10::DeviceType::CPU),
+          false);
+      fs.write(
+          (const char*)host_data.data() + tensor_offset, (long)tensor_nbyte);
```
This whole logic feels wrong (and will likely fails ASAN)
You should copy with offset, because your vector is only of `tensor_nbytes` so if offset is not zero here, you'll be writing outside of tensor boundary

shengfukevin replied:
Forgot the add offset on GPU case. fixed.

malfet reviewed (changes requested):
This is almost guaranteed to trigger OOB when saving tensor with offset which is not on CPU

briancoutinho reviewed (approved):
lgtm:)

shengfukevin committed:
Save integral tensor data for ET (#144508)

Summary:

et_replay uses random data to run operators, however, the operators using index tensor to access memory won't work with random data. It usually ran into two exceptions: 1. illegal memory access since index is out of range, it has been fixed with the environment variable ENABLE_PYTORCH_EXECUTION_TRACE_SAVE_INTEGRAL_TENSOR_RANGE to record the min/max value of index tensors. 2. unaligned memory access, FBGEMM ops have speical requirements for the memory layout.

To fix the second execption, ENABLE_PYTORCH_EXECUTION_TRACE_SAVE_INTEGRAL_TENSOR is added to allow user to specify the node names, separated by comma, so ET will save the integral tensor data for these nodes. The saved data will be used in et_replay.

Be careful to turn on this option since it will use more space to save the extra data.

Test Plan: buck2 run mode/opt caffe2/test:test_profiler_cuda -- profiler.test_execution_trace.TestExecutionTraceCUDA.test_execution_trace_record_integral_tensor_data_cuda

Reviewed By: briancoutinho

Differential Revision: D67989856
